src/modules/emit_sound.rs

```diff
@@ -0,0 +1,229 @@
+            && engine::S_PrecacheSound.is_set(marker)
+            && engine::listener_origin.is_set(marker)
+    }
+}
+
+struct SoundInfo {
+    sound: String,
+    channel: i32,
+    /// Volume is [0..1]
+    volume: f32,
+    from: i32,
+    to: i32,
+    /// This is the "radius" of the sound. Close to 1, you hear sound quieter when moving farther
+    /// from entity. "NORM" for sound-emitting entity is 0.8.
+    /// 0 is not quite to make it "global". Negative value will do.
+    attenuation: f32,
+    flag: i32,
+    /// 100 is no shift.
+    pitch: i32,
+}
+
+impl SoundInfo {
+    fn new(sound: String, channel: i32) -> Self {
+        SoundInfo {
+            sound,
+            channel,
+            volume: 1.,
+            from: 0,
+            to: 0,
+            attenuation: 0.,
+            flag: 0,
+            pitch: 100,
+        }
+    }
+}
+
+// It goes up to 14-15 something. I've seen up to 8 used so far.
+// There should be only 8 usable channels from SV_StartSounds.
+// But SND_PickDynamicChannel says something very weird otherwise.
+static MAX_CHANNELS: i32 = 10;
+
+// Eh.
+// Parse float then round then convert to integer to avoid a more convoluted solution.
+impl FromStr for SoundInfo {
+    type Err = ParseFloatError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.split_ascii_whitespace();
+        let sound = iter.next().unwrap_or("").to_string();
+        let channel =
+            ((f32::from_str(iter.next().unwrap_or(""))?).round() as i32).clamp(0, MAX_CHANNELS);
+        let volume = (f32::from_str(iter.next().unwrap_or(""))?).clamp(0., 1.);
+        let from = ((f32::from_str(iter.next().unwrap_or(""))?).round() as i32).clamp(0, 2047);
+        let to = ((f32::from_str(iter.next().unwrap_or(""))?).round() as i32).clamp(0, 2047);
+        let attenuation = (f32::from_str(iter.next().unwrap_or(""))?).clamp(-1., 1.);
+        let flag = ((f32::from_str(iter.next().unwrap_or(""))?).round() as i32).clamp(0, 255);
+        let pitch = ((f32::from_str(iter.next().unwrap_or(""))?).round() as i32).clamp(0, 255);
+
+        Ok(SoundInfo {
+            sound,
+            channel,
+            volume,
+            from,
+            to,
+            attenuation,
+            flag,
+            pitch,
+        })
+    }
+}
+
+static BXT_EMIT_SOUND: Command = Command::new(
+    b"bxt_emit_sound\0",
+    handler!(
+        "bxt_emit_sound <sound> <channel> [volume] [from] [to] [attenuation] [flag] [pitch]
+Plays sound file directly from SV_StartSound.",
+        emit_sound as fn(_, _, _),
+        emit_sound_full as fn(_, _)
+    ),
+);
+
+fn emit_sound(marker: MainThreadMarker, sound: String, channel: i32) {
+    emit_sound_full(marker, SoundInfo::new(sound, channel));
+}
+
+fn emit_sound_full(marker: MainThreadMarker, info: SoundInfo) {
+    if let Some(player) = unsafe { engine::player_edict(marker) } {
+        let entity = unsafe { player.as_ptr().add(info.from as usize).as_mut() };
```

Comment:
This should not use the player edict pointer as the base. It should use `server_t->edicts` and check `server_t->num_edicts` before accessing the pointer, as the edict count is dynamically allocated (and not statically bounded by 2047 as the `clamp`s above seem to imply). Something like this in BXT: https://github.com/YaLTeR/BunnymodXT/blob/71f1c9a75d6e92cb1f5e48c8d380541451165ea8/BunnymodXT/modules/HwDLL.hpp#L227-L230